File: source/2018/day_01/C#/Puzzle_2.cs
using System;
using System.Collections.Generic;

namespace Day_01
{
    class Puzzle_2
    {
        public static int Solve(int[] input)
        {
            int result = 0;
            //Declares a HashSet to save unique elements
            HashSet<int> frequencies = new HashSet<int>();

            while (true)
            {
                foreach (int number in input)
                {
                    result += number;

                    //checks if current sum(result) already exists inside the HashSet and if not to add
                    if (!frequencies.Add(result))
                    {
                        //Breaks out of all loops (and function) and returns the result
                        return result;
                    }
                }
            }
        }
    }
}

File: source/2018/day_01/Python/puzzle_2.py
def solve(data):
    result = 0
    # Creates and unordered collection of unique elements
    unique_frequencies = set()

    while True:
        for number in data:
            result += number

            # Checks if current sum(result) already exists in the collection
            if result in unique_frequencies:
                # Breaks out of the function and returns the result
                return result

            # Adds current sum(result) to the collection
            unique_frequencies.add(result)